audit/logger.go

```diff
@@ -0,0 +1,161 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package audit
+
+import (
+	"strconv"
+	"github.com/control-center/serviced/datastore"
+	"github.com/control-center/serviced/logging"
+	"github.com/Sirupsen/logrus"
+	"github.com/zenoss/logri"
+)
+
+var plog = logging.PackageLogger()
+
+// Logger is the interface for audit logging.  Any implementations for
+// audit logging should implement this interface.
+type Logger interface {
+
+	// Set the action that we are auditing.
+	Action(action string) Logger
+
+	// Set the message that we are writing to the audit log.
+	Message(ctx datastore.Context, message string) Logger
+
+	// Set the type of entity being modified.
+	Type(theType string) Logger
+
+	// Set the id of the entity being modified.
+	ID(id string) Logger
+
+	// Set the type of entity being modified.
+	Entity(entity datastore.Entity) Logger
+
+	// Add an additional field to the entry.
+	WithField(name string, value string) Logger
+
+	// Add additional fields to the entry.
+	WithFields(fields logrus.Fields) Logger
+
+	// Log that the action succeeded.
+	Succeeded()
+
+	// Log that the action failed.
+	Failed()
+
+	// Log whether the action succeeded or failed based on the value passed in.
+	SucceededIf(value bool)
+
+	// Log whether the action succeeded or failed based on the error passed in.
+	Error(err error) error
+}
+
+// NewLogger returns a default implmentation of the audit logger.  The "user" will default
+// to "system" unless otherwise specified in the context.  This wraps a logri Logger,
+// and will write to the location specified in the logger config.
+func NewLogger() Logger {
+	l := logri.GetLogger("audit")
+	return &logger{loggeri: l}
+}
+
+type logger struct {
+	entry   *logrus.Entry
+	message string
+	loggeri *logri.Logger
+}
+
+func (l *logger) Action(action string) Logger {
+	return l.addField("action", action)
+}
+
+func (l *logger) Message(ctx datastore.Context, message string) Logger {
+    l.addField("user", ctx.User())
+    l.message = message
+    return l
+}
+
+func (l *logger) Type(theType string) Logger {
+	return l.addField("type", theType)
+}
+
+func (l *logger) ID(id string) Logger {
+	return l.addField("id", id)
+}
+
+func (l *logger) Entity(entity datastore.Entity) Logger {
+	return l.addFields(logrus.Fields{
+		"id":   entity.GetID(),
+		"type": entity.GetType(),
+		})
+}
+
+func (l *logger) WithField(name string, value string) Logger {
+    return l.addField(name, value)
+}
+
+func (l *logger) addField(name string, value string) Logger {
+    if l.entry != nil {
+        l.entry = l.entry.WithField(name, value)
+    } else {
+        l.entry = l.loggeri.WithField(name, value)
+        l.entry.Logger.Hooks = make(map[logrus.Level][]logrus.Hook)
+    }
+    return l
```

Comment:
Sorry I missed this when I first suggested, but this could a little DRYer by having `addField` just call `addFields` with a list of one item:
```
    return l.addFields(logrus.Fields{name, value})
```